controllers/awscluster_controller.go

```diff
@@ -66,48 +66,83 @@ var (
+	elbServiceFactory     awsClusterELBServiceFactory
+	securityGroupFactory  awsClusterSecurityGroupServiceFactory
 	Endpoints             []scope.ServiceEndpoint
 	WatchFilterValue      string
 }
 
-// getEC2Service factory func is added for testing purpose so that we can inject mocked EC2Service to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getEC2Service(scope scope.EC2Scope) services.EC2Interface {
-	if r.ec2ServiceFactory != nil {
-		return r.ec2ServiceFactory(scope)
+type NewClusterReconcilerInput struct {
+	Manager          ctrl.Manager
+	WatchFilterValue string
+	Endpoints        []scope.ServiceEndpoint
+}
+
+type awsClusterEC2ServiceFactory func(scope.EC2Scope) services.EC2Interface
+
+func withAWSClusterEC2ServiceFactory(fn awsClusterEC2ServiceFactory) awsClusterReconcilerOption {
+	return func(r *AWSClusterReconciler) {
+		r.ec2ServiceFactory = fn
 	}
-	return ec2.NewService(scope)
 }
 
-// getELBService factory func is added for testing purpose so that we can inject mocked ELBService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getELBService(scope scope.ELBScope) services.ELBInterface {
-	if r.elbServiceFactory != nil {
-		return r.elbServiceFactory(scope)
+type awsClusterELBServiceFactory func(scope.ELBScope) services.ELBInterface
+
+func withAWSClusterELBServiceFactory(fn awsClusterELBServiceFactory) awsClusterReconcilerOption {
+	return func(r *AWSClusterReconciler) {
+		r.elbServiceFactory = fn
 	}
-	return elb.NewService(scope)
 }
 
-// getNetworkService factory func is added for testing purpose so that we can inject mocked NetworkService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getNetworkService(scope scope.ClusterScope) services.NetworkInterface {
-	if r.networkServiceFactory != nil {
-		return r.networkServiceFactory(scope)
+type awsClusterNetworkServiceFactory func(scope.ClusterScope) services.NetworkInterface
+
+func withAWSClusterNetworkServiceFactory(fn awsClusterNetworkServiceFactory) awsClusterReconcilerOption {
+	return func(r *AWSClusterReconciler) {
+		r.networkServiceFactory = fn
 	}
-	return network.NewService(&scope)
 }
 
-// getSecurityGroupService factory func is added for testing purpose so that we can inject mocked SecurityGroupService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getSecurityGroupService(scope scope.ClusterScope) services.SecurityGroupInterface {
-	if r.securityGroupFactory != nil {
-		return r.securityGroupFactory(scope)
+type awsClusterSecurityGroupServiceFactory func(scope.ClusterScope) services.SecurityGroupInterface
+
+func withAWSClusterSecurityGroupServiceFactory(fn awsClusterSecurityGroupServiceFactory) awsClusterReconcilerOption {
+	return func(r *AWSClusterReconciler) {
+		r.securityGroupFactory = fn
 	}
-	return securitygroup.NewService(&scope, awsSecurityGroupRoles)
+}
+
+func NewClusterReconciler(input NewClusterReconcilerInput, opt ...awsClusterReconcilerOption) *AWSClusterReconciler {
```

Comment:
What do you think to `AWSClusterReconciler` being changed so that is not exportable (i.e. `awsClusterReconciler`) and that we have an interface that is returned by `NewClusterReconciler`

Reply:
Sounds good to me, will give it a try.